packages/gatsby/index.d.ts

```diff
@@ -640,69 +681,142 @@ type GatsbyStages =
+  extends ParentSpanPluginArgs<TSource, TContext> {
   app: any
 }
 
-export interface CreateNodeArgs extends ParentSpanPluginArgs {
-  node: Node
+export interface CreateNodeArgs<
+  TNode extends object = {},
+  TSource = any,
+  TContext = any
+> extends ParentSpanPluginArgs<TSource, TContext> {
+  node: Node & TNode
   traceId: string
   traceTags: {
     nodeId: string
     nodeType: string
   }
 }
 
-export interface CreatePageArgs extends ParentSpanPluginArgs {
-  page: Node
+export interface CreatePageArgs<
+  TNode extends object = {},
+  TSource = any,
+  TContext = any
+> extends ParentSpanPluginArgs<TSource, TContext> {
+  page: Node & TNode
   traceId: string
 }
 
-export interface CreateWebpackConfigArgs extends ParentSpanPluginArgs {
+export interface CreateWebpackConfigArgs<TSource = any, TContext = any>
+  extends ParentSpanPluginArgs<TSource, TContext> {
   getConfig: Function
   stage: GatsbyStages
   rules: WebpackRules
   loaders: WebpackLoaders
   plugins: WebpackPlugins
 }
 
-export interface PreprocessSourceArgs extends ParentSpanPluginArgs {
+export interface PreprocessSourceArgs<TSource = any, TContext = any>
+  extends ParentSpanPluginArgs<TSource, TContext> {
   filename: string
   contents: string
 }
 
-export interface ResolvableExtensionsArgs extends ParentSpanPluginArgs {
+export interface ResolvableExtensionsArgs<TSource = any, TContext = any>
+  extends ParentSpanPluginArgs<TSource, TContext> {
   traceId: "initial-resolvableExtensions"
 }
 
-export interface SetFieldsOnGraphQLNodeTypeArgs extends ParentSpanPluginArgs {
+export interface SetFieldsOnGraphQLNodeTypeArgs<TSource = any, TContext = any>
+  extends ParentSpanPluginArgs<TSource, TContext> {
   type: {
     name: string
     nodes: any[]
   }
   traceId: "initial-setFieldsOnGraphQLNodeType"
 }
 
-export interface SourceNodesArgs extends ParentSpanPluginArgs {
+export interface GatsbyGraphQLObjectType<TSource = any, TContext = any> {
+  kind: "OBJECT"
+  config: ComposeObjectTypeConfig<TSource, TContext>
+}
+
+interface GatsbyGraphQLInputObjectType {
+  kind: "INPUT_OBJECT"
+  config: ComposeInputObjectTypeConfig
+}
+
+interface GatsbyGraphQLUnionType<TSource = any, TContext = any> {
+  kind: "UNION"
+  config: ComposeUnionTypeConfig<TSource, TContext>
+}
+
+interface GatsbyGraphQLInterfaceType<TSource = any, TContext = any> {
+  kind: "INTERFACE"
+  config: ComposeInterfaceTypeConfig<TSource, TContext>
+}
+
+interface GatsbyGraphQLEnumType {
+  kind: "ENUM"
+  config: ComposeEnumTypeConfig
+}
+
+interface GatsbyGraphQLScalarType {
+  kind: "SCALAR"
+  config: ComposeScalarTypeConfig
+}
+
+export type GatsbyGraphQLType<TSource = any, TContext = any> =
+  | GatsbyGraphQLObjectType
+  | GatsbyGraphQLInputObjectType
+  | GatsbyGraphQLUnionType<TSource, TContext>
+  | GatsbyGraphQLInterfaceType<TSource, TContext>
+  | GatsbyGraphQLEnumType
+  | GatsbyGraphQLScalarType
+
+export interface NodePluginSchema<TSource = any, TContext = any> {
```

Comment:
Is there value in those generics? Would user ever provide their own here? 

If so, we probably would want better names for them, because it isn't immediately apparent what `TSource` or `TContext` refer to when working with (for example) `sourceNodes` (specifically if you user doesn't touch schema stuff in their implementation)

If there is no real value, then I would vote hardcoding `any` for those things we import from `graphql`/`graphql-compose` - simple is good, and I think it's good to avoid generics (if not needed).

/cc @blainekasten Any thoughts on this?

Reply:
Turns out I'm really not using the generics much in my own [codebase](https://github.com/ifiokjr/ifiokjr.com/blob/3bf0cfb5b841b1725722b6cbd5164c6031b1e46d/src/gatsby/source-nodes.ts#L28) so I'll push up new commit without them.